openstack/baremetal/v1/nodes/requests.go

```diff
@@ -478,3 +479,89 @@ func ChangePowerState(client *gophercloud.ServiceClient, id string, opts PowerSt
 	})
 	return
 }
+
+// This is the desired RAID configuration on the bare metal node.
+type TargetRAIDConfig struct {
+	LogicalDisks []LogicalDisk `json:"logical_disks"`
+}
+
+// TargetRAIDConfigBuilder allows extensions to modify a set RAID config request.
+type TargetRAIDConfigBuilder interface {
+	ToRAIDConfigMap() (map[string]interface{}, error)
+}
+
+type LogicalDisk struct {
+	// Size (Integer) of the logical disk to be created in GiB.  If unspecified, "MAX" will be used.
+	SizeGB *int `json:"size_gb"`
+
+	// RAID level for the logical disk.
+	RAIDLevel string `json:"raid_level" required:"true"`
+
+	// Name of the volume. Should be unique within the Node. If not specified, volume name will be auto-generated.
+	VolumeName string `json:"volume_name,omitempty"`
+
+	// Set to true if this is the root volume. At most one logical disk can have this set to true.
+	IsRootVolume *bool `json:"is_root_volume,omitempty"`
+
+	// Set to true if this logical disk can share physical disks with other logical disks.
+	SharePhysicalDisks *bool `json:"share_physical_disks,omitempty"`
+
+	// If this is not specified, disk type will not be a criterion to find backing physical disks
+	DiskType string `json:"disk_type,omitempty"`
+
+	// If this is not specified, interface type will not be a criterion to find backing physical disks.
+	InterfaceType string `json:"interface_type,omitempty"`
+
+	// Integer, number of disks to use for the logical disk. Defaults to minimum number of disks required
+	// for the particular RAID level.
+	NumberOfPhysicalDisks int `json:"number_of_physical_disks,omitempty"`
+
+	// The name of the controller as read by the RAID interface.
+	Controller string `json:"controller,omitempty"`
+
+	// A list of physical disks to use as read by the RAID interface.
+	PhysicalDisks []map[string]interface{} `json:"physical_disks,omitempty"`
+}
+
+// Custom marshal function for LogicalDisk to send "MAX" if SizeGB isn't specified.
+func (disk *LogicalDisk) MarshalJSON() ([]byte, error) {
+	type Alias LogicalDisk
+
+	var sizeGB interface{}
+	if disk.SizeGB == nil {
+		sizeGB = "MAX"
+	} else {
+		sizeGB = disk.SizeGB
+	}
+
+	return json.Marshal(&struct {
+		SizeGB interface{} `json:"size_gb"`
+		*Alias
+	}{
+		SizeGB: sizeGB,
+		Alias:  (*Alias)(disk),
+	})
+}
+
+func (t *TargetRAIDConfig) ToRAIDConfigMap() (map[string]interface{}, error) {
```

Comment:
Instead of implementing `MarshalJSON`, we can add the logic here:

```go
func (opts RAIDConfigOpts) ToRAIDConfigMap() (map[string]interface{}, error) {
  body, err := gophercloud.BuildRequestBody(opts, "")
  if err != nil {
    return nil, err
  }

  for _, v := range body["logical_disks"].([]interface{}) {
    if logicalDisk, ok := v.(map[string]interface{}); ok {
      if logicalDisk["size_gb"] == nil {
        logicalDisk["size_gb"] = "MAX"
      }
    }
  }

  return body, nil
}
```

There is little difference between `MarshalJSON` and the `ToXYMap()` methods. I'm running on little sleep today, so my articulation is low, but one of the key differences has to do with the ability to extend the various "opts" structs, for example: https://github.com/gophercloud/gophercloud/blob/master/openstack/compute/v2/extensions/keypairs/requests.go#L9-L32

A minor difference is simply that since we're defining a `ToXYMap()` method, implementing `MarshalJSON` is slightly redundant since the method can do the same thing.

If you weren't implementing an interface, then `MarshalJSON` would be the only way to achieve this logic.

Reply:
Thanks, I think that looks a lot nicer